### PR TITLE
fix chat-link in end of line displays tooltip over text and not on link

### PR DESCRIPTION
**Pull request by teneeto**
### Details

This PR seeks to address the reviews of a previously authored and [closed PR](https://github.com/Expensify/App/pull/27817)
Referencing the leading solution https://github.com/Expensify/App/pull/27817#issue-1903838307

Add a new prop to Tooltip called shouldUseMultilinePositioning. This enables a new algorithm for finding the bounding box target for the tooltip. In the case where the target has wrapped onto multiple lines (e.g. it is a link in a chat window) then the link has multiple bounding boxes, and we want to show the tooltip against the one that the user is hovering over.

As part of this, extend Hoverable to pass the Event to its onHoverIn / onHoverOut callbacks.

Enable this new algorithm from BaseAnchorForCommentsOnly, which is the base class for links that show up in chat.

### Fixed Issues

$ https://github.com/Expensify/App/issues/27585
PROPOSAL:  https://github.com/Expensify/App/issues/27585

### Tests

Referencing the leading solution https://github.com/Expensify/App/pull/27817#issue-1903838307

Action Performed:

On a platform with a mouse (i.e. not mobile):

Open the app
Open any report
Type in some text and at the end of line, add link with long text eg: Test test test test test test test test test test
Send the text, text should have text and link in first line and link should be continued in second line too. Adjust the window size to force the text to wrap if necessary.
Hover on link and observe that tooltip is displayed over text
Expected Result:
App should display tooltip over link on hover. The tooltip should appear over the same section of the link that the user is hovering over.

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/22779249/ac7552eb-63c0-4cc6-b226-48b12ee1c736

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

File: src/components/Tooltip/index.js [hunk 1]
```diff
@@ -1,37 +1,239 @@
...
     );
 }
 
-Tooltip.displayName = 'Tooltip';
-Tooltip.propTypes = propTypes;
-Tooltip.defaultProps = defaultProps;
-
-export default Tooltip;
+Tooltip.propTypes = tooltipPropTypes.propTypes;
+Tooltip.defaultProps = tooltipPropTypes.defaultProps;
```
Comment (gedu) [inline src/components/Tooltip/index.js:238]: why did you remove the `displayName`?

File: src/components/Tooltip/index.js [hunk 2]
```diff
@@ -1,37 +1,239 @@
...
+}
+
+/**
+ * A component used to wrap an element intended for displaying a tooltip. The term "tooltip's target" refers to the
+ * wrapped element, which, upon hover, triggers the tooltip to be shown.
+ * @param {propTypes} props
+ * @returns {ReactNodeLike}
+ */
+function Tooltip(props) {
+    const {children, numberOfLines, maxWidth, text, renderTooltipContent, renderTooltipContentKey} = props;
```
Comment (BeeMargarida) [inline src/components/Tooltip/index.js:80]: Can't this be deconstructed above?

File: src/components/Tooltip/index.js [hunk 3]
```diff
@@ -1,37 +1,239 @@
...
+    const [isRendered, setIsRendered] = useState(false);
+    // Is the tooltip currently visible?
+    const [isVisible, setIsVisible] = useState(false);
+    // The distance between the left side of the wrapper view and the left side of the window
+    const [xOffset, setXOffset] = useState(0);
+    // The distance between the top of the wrapper view and the top of the window
+    const [yOffset, setYOffset] = useState(0);
+    // The width and height of the wrapper view
+    const [wrapperWidth, setWrapperWidth] = useState(0);
+    const [wrapperHeight, setWrapperHeight] = useState(0);
```
Comment (BeeMargarida) [inline src/components/Tooltip/index.js:95]: The height and width could be joined in a object, so there's less calls to update since they always update together. The same for the offsets.

File: src/components/Tooltip/index.js [hunk 4]
```diff
@@ -1,37 +1,239 @@
...
+            )}
+            <BoundsObserver
+                enabled={isVisible}
+                onBoundsChange={updateBounds}
+            >
+                <Hoverable
+                    onMouseEnter={updateTargetAndMousePosition}
+                    onHoverIn={showTooltip}
+                    onHoverOut={hideTooltip}
+                    shouldHandleScroll={props.shouldHandleScroll}
```
Comment (BeeMargarida) [inline src/components/Tooltip/index.js:228]: can't this prop be deconstructed above?

File: src/components/Tooltip/index.js (hunk 1 above)
Comment (teneeto) [reply src/components/Tooltip/index.js:238]: O will have it back

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/components/Hoverable/index.js [hunk 5]
```diff
@@ -46,6 +46,7 @@ class Hoverable extends Component {
                      * If the user has started scrolling and the isHoveredRef is true, then we should set the hover state to false.
                      * This is to hide the existing hover and reaction bar.
                      */
+                    this.isHoveredRef = false;
```
[the remark this answers is not in the document]
Comment (teneeto) [reply src/components/Hoverable/index.js:49]: On a second look, it's really not necessary.

File: src/components/Tooltip/index.js (hunk 3 above)
Comment (teneeto) [reply src/components/Tooltip/index.js]: This code here has been moved to base tooltip, which already has the logics there.

File: src/components/Tooltip/index.js (hunk 2 above)
Comment (teneeto) [reply src/components/Tooltip/index.js]: It can but props was used with underscore. Ideally I think it's fine this way for now.

**Comment by teneeto** [comment]
Recheck

File: src/components/Tooltip/BaseTooltip.js [hunk 6]
```diff
@@ -19,8 +19,42 @@ const hasHoverSupport = DeviceCapabilities.hasHoverSupport();
...
+
+    // If no matching bounding box is found, fall back to the first one.
+    // This could only happen if the user is moving the mouse very quickly
+    // and they got it outside our slop above.
+    return bbs[0];
+}
+
 function Tooltip(props) {
-    const {children, numberOfLines, maxWidth, text, renderTooltipContent, renderTooltipContentKey} = props;
+    const {children, numberOfLines, maxWidth, text, renderTooltipContent, renderTooltipContentKey, shouldHandleScroll} = props;
```
Comment (s77rt) [inline src/components/Tooltip/BaseTooltip.js:57]: Move this to the function parameter 

```js
function Tooltip({children, numberOfLines, maxWidth, text, renderTooltipContent, renderTooltipContentKey, shouldHandleScroll}) {
```

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/components/Hoverable/index.js [hunk 7]
```diff
@@ -42,10 +42,6 @@ class Hoverable extends Component {
                 if (!scrolling && this.isHoveredRef) {
                     this.setState({isHovered: this.isHoveredRef}, this.props.onHoverIn);
                 } else if (scrolling && this.isHoveredRef) {
-                    /**
-                     * If the user has started scrolling and the isHoveredRef is true, then we should set the hover state to false.
-                     * This is to hide the existing hover and reaction bar.
-                     */
```
[the remark this answers is not in the document]
Comment (teneeto) [reply src/components/Hoverable/index.js:48]: you mean just the comments or the comment and the initial code `this.isHoveredRef = false;`

File: src/components/Tooltip/BaseTooltip.js (hunk 6 above)
Comment (teneeto) [reply src/components/Tooltip/BaseTooltip.js:57]: We have two lines of code using props and underscore.
`shiftHorizontal={_.result(props, 'shiftHorizontal')}`
`shiftVertical={_.result(props, 'shiftVertical')}`


Should I remove the use of underscore and pass them directly?

or could we have something like this?
```function Tooltip(props, {children, numberOfLines, maxWidth, text, renderTooltipContent, renderTooltipContentKey, shouldHandleScroll}) {```

File: src/components/Hoverable/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (s77rt) [reply src/components/Hoverable/index.js:48]: Just the comments. Why we are deleting those comments?

File: src/components/Tooltip/BaseTooltip.js (hunk 6 above)
Comment (s77rt) [reply src/components/Tooltip/BaseTooltip.js:57]: For those two props use `Str.result` (from expensify-common)

```js
shiftHorizontal={Str.result(shiftHorizontal)}
shiftVertical={Str.result(shiftHorizontal)}
```

File: src/components/Hoverable/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (teneeto) [reply src/components/Hoverable/index.js:48]: comm

> Just the comments. Why we are deleting those comments?

comments were related to this `this.isHoveredRef = false;`. So I removed them since I was taking off that piece of code.

[the remark this answers is not in the document]
Comment (s77rt) [reply src/components/Hoverable/index.js:48]: In the comment we are explaining why we are setting the hover state to false and we are still setting it to false.

[the remark this answers is not in the document]
Comment (teneeto) [reply src/components/Hoverable/index.js:48]: ok, I will revert.

**Comment by teneeto** [comment]
@s77rt It's all resolved, you can recheck. Thanks.

**Comment by s77rt** [comment]
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I verified tests pass on **all platforms** & I tested again on:
    - [X] Android: Native
    - [X] Android: mWeb Chrome
    - [X] iOS: Native
    - [X] iOS: mWeb Safari
    - [X] MacOS: Chrome / Safari
    - [X] MacOS: Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [X] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>




https://github.com/Expensify/App/assets/16493223/3942a069-3f84-4157-a499-581275c6250d


</details>

<details>
<summary>Mobile Web - Chrome</summary>



</details>

<details>
<summary>Mobile Web - Safari</summary>



</details>

<details>
<summary>Desktop</summary>




https://github.com/Expensify/App/assets/16493223/b47b4d96-241c-4ccb-a49a-7094c641f8e8



</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Review by s77rt — APPROVED** [review]

**Comment by s77rt** [comment]
@teneeto Thanks!

**Review by pecanoro — APPROVED** [review]

File: src/components/Tooltip/BaseTooltip.js [hunk 8]
```diff
@@ -91,10 +132,15 @@ function Tooltip(props) {
...
-        setWrapperWidth(bounds.width);
-        setWrapperHeight(bounds.height);
-        setXOffset(bounds.x);
-        setYOffset(bounds.y);
+        // Choose a bounding box for the tooltip to target.
+        // In the case when the target is a link that has wrapped onto
+        // multiple lines, we want to show the tooltip over the part
+        // of the link that the user is hovering over.
+        const betterBounds = chooseBoundingBox(target.current, initialMousePosition.current.x, initialMousePosition.current.y);
+        setWrapperWidth(betterBounds.width);
```
Comment (tgolen) [inline src/components/Tooltip/BaseTooltip.js:140]: I believe this is leading to a crash with `Uncaught TypeError: Cannot read properties of undefined (reading 'width')`. This is what I did:

1. Sign in with a new account
2. Go to Settings > Workspaces > Create New Workspace (don't create anything)
3. Click the back button on the modal
image


image

Comment (s77rt) [reply src/components/Tooltip/BaseTooltip.js:140]: Yes it's, more info https://github.com/Expensify/App/issues/29446#issuecomment-1759942603 and PR fixing it https://github.com/Expensify/App/pull/29468

File: src/components/Tooltip/BaseTooltip.js [hunk 9]
```diff
@@ -43,6 +76,14 @@ function Tooltip(props) {
     const isAnimationCanceled = useRef(false);
     const prevText = usePrevious(text);
 
+    const target = useRef(null);
+    const initialMousePosition = useRef({x: 0, y: 0});
+
+    const updateTargetAndMousePosition = useCallback((e) => {
+        target.current = e.target;
```
Comment (s77rt) [inline src/components/Tooltip/BaseTooltip.js:83]: We should have used the `currentTarget` (the element that is listening to the mouse enter event) instead of `target` (the element that caused the event to fire and bubble up). Coming from https://github.com/Expensify/App/issues/29678
